clang/lib/Sema/SemaStmtAttr.cpp

```diff
@@ -322,6 +322,56 @@ static Attr *handleUnlikely(Sema &S, Stmt *St, const ParsedAttr &A,
   return ::new (S.Context) UnlikelyAttr(S.Context, A);
 }
 
+CodeAlignAttr *Sema::BuildCodeAlignAttr(const AttributeCommonInfo &CI,
+                                        Expr *E) {
+  if (!E->isValueDependent()) {
+    llvm::APSInt ArgVal;
+    ExprResult Res = VerifyIntegerConstantExpression(E, &ArgVal);
+    if (Res.isInvalid())
+      return nullptr;
+    E = Res.get();
+
+    // This attribute requires a strictly positive value.
+    if (ArgVal <= 0) {
+      Diag(E->getExprLoc(), diag::err_attribute_requires_positive_integer)
+          << CI << /*positive*/ 0;
+      return nullptr;
+    }
+
+    // This attribute requires a single constant power of two greater than zero.
+    if (!ArgVal.isPowerOf2()) {
+      Diag(E->getExprLoc(), diag::err_attribute_argument_not_power_of_two)
+          << CI;
+      return nullptr;
+    }
+  }
+
+  return new (Context) CodeAlignAttr(Context, CI, E);
+}
+
+static Attr *handleCodeAlignAttr(Sema &S, Stmt *St, const ParsedAttr &A) {
+
+  Expr *E = A.getArgAsExpr(0);
+  return S.BuildCodeAlignAttr(A, E);
+}
+
+// Emit duplicate error for [[clang::code_align()]] attribute.
+template <typename LoopAttrT>
+static void
+CheckForDuplicateLoopAttribute(Sema &S,
+                               const SmallVectorImpl<const Attr *> &Attrs) {
+  const LoopAttrT *LoopAttr = nullptr;
+
+  for (const auto *I : Attrs) {
+    if (LoopAttr && isa<LoopAttrT>(I)) {
+      // Cannot specify same type of attribute twice.
+      S.Diag(I->getLocation(), diag::err_loop_attr_duplication) << LoopAttr;
+    }
+    if (isa<LoopAttrT>(I))
+      LoopAttr = cast<LoopAttrT>(I);
+  }
+}
+
```

Comment:
It's probably a bit hostile for it to be an error if the values are the same, but also, I'm not certain how you'd get duplicate attributes on a statement without a copy-paste error so maybe this is fine.
```suggestion
CheckForDuplicateLoopAttribute(Sema &S,
                               const SmallVectorImpl<const Attr *> &Attrs) {
  const Attr *A = nullptr;

  for (const auto *I : Attrs) {
    if (isa<LoopAttrT>(I)) {
      if (A) {
        // Cannot specify same type of attribute twice.
        S.Diag(I->getLocation(), diag::err_loop_attr_duplication) << A;
       }
       A = I;
    }
  }
}

```

Reply:
I could definitely see macros causing duplicates, presumably as we get ones that are compound/for different functions.  I would think ones that have the same value aren't worth diagnosing on.

Reply:
I have updates codes for duplicate attribute values check https://github.com/llvm/llvm-project/pull/70762/commits/d049dc9997bdb78ff7e7cbf9b04aa42b9274cfd9